lib/tree.js

```diff
@@ -12,40 +12,124 @@ class Tree {
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    // Time Complexity: O(log(n)) - if a balanced tree, otherwise, O(n)
+    // Space Complexity: O(1)
     find(key) {
-        throw new Error("This method hasn't been implemented yet!");
+        if (!this.root) return null;
+        
+        let current = this.root;
+
+        while (current) {
+            if (current.key === key) {
+                return current.value;
+            } else if (key < current.key) {
+                current = current.left;
+            } else {
+                current = current.right;
+            }
+        }
+
+        return null; // if key is not found
+    }
+
+    // Time Complexity: O(n) - touching every node
+    // Space Complexity: O(n) - array is length n, and n number of recursive calls on the stack
+    // inorder: left, current, right (left most nodes get pushed to list first, right last since it's largest)
+    inorderHelper(current, array) {
+        if (current) {
+            this.inorderHelper(current.left, array);
+            array.push({key: current.key, value: current.value});
+            this.inorderHelper(current.right, array);
+        }
+
+        return array;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
     inorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.inorderHelper(this.root, [])
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
+    // preorder: current, left, right (parent nodes get pushed to list first)
+    preorderHelper(current, array) {
+        if (current) {
+            array.push({key: current.key, value: current.value});
+            this.preorderHelper(current.left, array);
+            this.preorderHelper(current.right, array);
+        }
+        
+        return array;
+    }
+    
+    
     preorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.preorderHelper(this.root, [])
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
+    // Postorder: left, right, current (get children nodes get pushed to list first)
+    postorderHelper(current, array) {
+        if (current) {
+            this.postorderHelper(current.left, array);
+            this.postorderHelper(current.right, array);
+            array.push({key: current.key, value: current.value});
+        }
+
+        return array;
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
     postorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.postorderHelper(this.root, [])
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
-    height() {
-        throw new Error("This method hasn't been implemented yet!");
+    // Time Complexity: O(n) - touches every node
+    // Space Complexity: O(log(n)) - for the call stack? 
+    height(current=this.root, count=0) {
```

Comment:
🤔 I don't see where this is evaluating which is larger the left or right subtree.  It seems to just return the count if the left subtree exists.